### PR TITLE
pin differently for build and run

### PR DESCRIPTION
**Pull request by 183amir**
Inspired by the discussion from here: https://github.com/conda-forge/staged-recipes/pull/1676#discussion_r81457522

@gqmelo:

> These compiled libraries seem to follow semantic version rules. So:
> -    they don't add any symbols when incrementing patch versions. That is why the recipe don't care exactly which version of xz is building against. Any patch version is equivalent in terms of ABI
> -    minor versions can add symbols (but not modify existing ones). That's why we need to pin xz and ncurses minor versions during build, but can allow to run with larger minor versions (they can only add things afterall).
> -    and major, well, they can just break everything, so the package should be totally rebuilt
> 
> For xz you can find an ABI report here: https://abi-laboratory.pro/tracker/timeline/xz/.
> Note that all versions are backward compatible, but from 5.0.8 to 5.2.0 they added 5 new symbols.
> 
> For ncurses I could not find such a report, but the release notes for 6.0 can give an idea: https://www.gnu.org/software/ncurses/
> 
> We could of course use the same pinning for running as we used for building here, but that would be too restrictive in terms of where the package could be installed and we would much more likely run into conflicts.

This should also fix the problems that I'm having on mac with `libpng` without needing to pin on a patch version.

### CODE REVIEW

**Comment by gqmelo** [comment]
Thanks for gathering all the ABI reports. I think this is very important to give users more flexibility in what package versions they can use.
It is really frustrating build a new version of some library just to find out that you will need to rebuild everything that depends on it.

File: scripts/pin_the_slow_way.py [hunk 1]
```diff
@@ -25,33 +25,80 @@
...
           'vlfeat': 'vlfeat 0.9.20',
           'xz': 'xz 5.2.*',
-          'zlib': 'zlib 1.2.*',
+          'zlib': 'zlib 1.2.7',
+        }
+
+run_pin = {
+          'boost': 'boost 1.61.*',
+          'bzip2': 'bzip2 >=1.0,<2',
+          'ffmpeg': 'ffmpeg >=2.8,<3.0',
```
Comment (gqmelo) [inline scripts/pin_the_slow_way.py]: According to the reports, they tend to break compatibility (medium severity) between minors. So I think it should be really ```2.8.*```

**Comment by patricksnape** [comment]
I'm fine with this although we'd need to make sure everyone was aware of it and that the ABI checks were done properly.

**Comment by 183amir** [comment]
I put the links for the places that I got my API/ABI compatibility reports and you cross check them. I think this pull is going to solve a couple of problems that we are having in the pinning department.

The idea is to build with the lowest possible version of libraries and set better runtime pinnings.

File: scripts/pin_the_slow_way.py [hunk 2]
```diff
@@ -25,33 +25,80 @@
...
+
+build_pin = {
           'boost': 'boost 1.61.*',
           'bzip2': 'bzip2 1.0.*',
           'ffmpeg': 'ffmpeg 2.8.*',
           'fontconfig': 'fontconfig 2.11.*',
           'freetype': 'freetype 2.6.*',
           'geos': 'geos 3.4.*',
-          'giflib': 'giflib 5.1.*',
+          'giflib': 'giflib 5.1.2',
```
Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: There is no link for `giflib`.

File: scripts/pin_the_slow_way.py [hunk 3]
```diff
@@ -25,33 +25,80 @@
...
           'hdf5': 'hdf5 1.8.17|1.8.17.*',
           'icu': 'icu 56.*',
           'jpeg': 'jpeg 9*',
           'libblitz': 'libblitz 0.10|0.10.*',
           'libmatio': 'libmatio 1.5.*',
-          'libnetcdf': 'libnetcdf 4.4.*',
```
Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: Please do not pin to the older version. There is a reason to use the latest one.

File: scripts/pin_the_slow_way.py [hunk 4]
```diff
@@ -25,33 +25,80 @@
...
-          'libpng': 'libpng >=1.6.21,<1.7',
+          'libnetcdf': 'libnetcdf 4.2.*',
+          'libpng': 'libpng 1.6.21',
           'libsvm': 'libsvm 3.21|3.21.*',
           'libtiff': 'libtiff 4.0.*',
           'ncurses': 'ncurses 5.9*',
           'openblas': 'openblas 0.2.18|0.2.18.*',
-          'openssl': 'openssl 1.0.*',
+          'openssl': 'openssl 1.0.2g',
```
Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: `conda-forge`'s version is `1.0.2,h` and that is 100% backward compatible.

File: scripts/pin_the_slow_way.py [hunk 5]
```diff
@@ -25,33 +25,80 @@
...
           'readline': 'readline 6.2*',
           'sox': 'sox 14.4.2',
-          'sqlite': 'sqlite 3.13.*',
+          'sqlite': 'sqlite 3.7.*',
```
Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: I wonder the reason for so many downgrades. This one will break a lot of stuff.

File: scripts/pin_the_slow_way.py [hunk 6]
```diff
@@ -25,33 +25,80 @@
...
           'tk': 'tk 8.5.*',
           'vlfeat': 'vlfeat 0.9.20',
           'xz': 'xz 5.2.*',
-          'zlib': 'zlib 1.2.*',
+          'zlib': 'zlib 1.2.7',
```
Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: Again, `conda-forge`'s version is 1.2.8 and that is `100%` backward compatible.

File: scripts/pin_the_slow_way.py (hunk 3 above)
Comment (183amir) [inline scripts/pin_the_slow_way.py]: Is just using the newer version at runtime not enough?

File: scripts/pin_the_slow_way.py (hunk 4 above)
Comment (183amir) [inline scripts/pin_the_slow_way.py]: The idea is to build with the lowest version that is forward compatible to some extent. This is a build constraint and during runtime packages will be run with the latest `1.0.*` version.

**Comment by 183amir** [comment]
@ocefpaf Most of the times using a newer version of library during runtime is enough to fix the problems as long as you can build with an old version but run against new versions.

**Comment by 183amir** [comment]
@ocefpaf I reverted the downgrades so that this is less controversial to go forward but in future the build dependencies should be as old as possible unless we have good reasons to use a newer version during build (not runtime).

File: scripts/pin_the_slow_way.py (hunk 2 above)
Comment (183amir) [inline scripts/pin_the_slow_way.py]: done

File: scripts/pin_the_slow_way.py (hunk 3 above)
Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: In this case no. Maybe not even in others.

File: scripts/pin_the_slow_way.py (hunk 2 above)
Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: And the link also says 100% backwards compatibility.

Comment (183amir) [inline scripts/pin_the_slow_way.py]: @ocefpaf I think you are mistaking what backward compatibility means here. `5.1.2` is 100% backwards compatible but that **does not mean** if you build with `5.1.2` (A new symbol was added), your compiled binary will work with `5.1.1`.
**It means** that if you build with `5.1.1`, your binary will work with `5.1.2` too.

What you are saying sounds like forward compatibility which I am not sure such thing even exists.

Comment (ocefpaf) [inline scripts/pin_the_slow_way.py]: That is exactly my understating. BTW that one new symbol may fail when tying to use that particular `symbol`. In general things would work in your scenario.

File: scripts/pin_the_slow_way.py [hunk 7]
```diff
@@ -25,38 +25,90 @@
...
-          'libpng': 'libpng >=1.6.23,<1.7',
+          'libpng': 'libpng 1.6.21',
           'libsvm': 'libsvm 3.21|3.21.*',
           'libtiff': 'libtiff 4.0.*',
           'ncurses': 'ncurses 5.9*',
           'netcdf-cxx4': 'netcdf-cxx4 4.3.*',
           'netcdf-fortran': 'netcdf-fortran 4.4.*',
           'openblas': 'openblas 0.2.18|0.2.18.*',
-          'openssl': 'openssl 1.0.*',
+          'openssl': 'openssl 1.0.2h',
```
Comment (jakirkham) [inline scripts/pin_the_slow_way.py]: So I don't think this level of pinning is needed for the runtime of `openssl`.

xref: https://www.openssl.org/docs/faq.html#MISC8

**Comment by jakirkham** [comment]
I think @insertinterestingnamehere had been proposing something similar in one of his recipes. Personally am open to this strategy in general. Not pinning our build versions exactly is probably causing us some suffering that is hard to determine (like when libraries break on the patch version) that this should help alleviate. There are some particular pins that I would like to discuss, but that is a separate discussion only worth getting into if we are in favor of the strategy as a whole.

File: scripts/pin_the_slow_way.py (hunk 2 above)
Comment (183amir) [inline scripts/pin_the_slow_way.py]: So is what I have done here good or not? Do you want me to change this to something else?

File: scripts/pin_the_slow_way.py [hunk 8]
```diff
@@ -25,38 +25,90 @@
...
-          'giflib': 'giflib 5.1.*',
+          'giflib': 'giflib 5.1.2',
           'hdf5': 'hdf5 1.8.17|1.8.17.*',
           'icu': 'icu 56.*',
           'jpeg': 'jpeg 9*',
           'libblitz': 'libblitz 0.10|0.10.*',
           'libmatio': 'libmatio 1.5.*',
           'libnetcdf': 'libnetcdf 4.4.*',
-          'libpng': 'libpng >=1.6.23,<1.7',
+          'libpng': 'libpng 1.6.21',
```
Comment (183amir) [inline scripts/pin_the_slow_way.py]: Are they needed during build?

File: scripts/pin_the_slow_way.py (hunk 7 above)
Comment (183amir) [inline scripts/pin_the_slow_way.py]: This is not a runtime pin, this is for building only.

File: scripts/pin_the_slow_way.py [hunk 9]
```diff
@@ -25,38 +25,90 @@
...
+          'jpeg': 'jpeg 9*',
+          'libblitz': 'libblitz 0.10|0.10.*',
+          'libmatio': 'libmatio 1.5.*',
+          'libnetcdf': 'libnetcdf >=4.4,<5',
+          'libpng': 'libpng >=1.6.21,<1.7',
+          'libsvm': 'libsvm 3.21|3.21.*',
+          'libtiff': 'libtiff >=4.0,<5',
+          'ncurses': 'ncurses >=5.9,<6',
+          'openblas': 'openblas 0.2.18|0.2.18.*',
+          'openssl': 'openssl >=1.0.2h,<1.1',
```
Comment (183amir) [inline scripts/pin_the_slow_way.py]: runtime pinnings are here @jakirkham

**Comment by ccordoba12** [comment]
It is an interesting idea because it seems to give more stability and robustness to the building process, and more flexibility to the running one.

I'm +1 on the idea, although I don't have time to review the implementation :-)

**Comment by 183amir** [comment]
@conda-forge/core this requires your attention on the idea behind this pull request. Thank you.

File: scripts/pin_the_slow_way.py (hunk 7 above)
Comment (jakirkham) [inline scripts/pin_the_slow_way.py]: Yeah, when I made that comment it was in the run section. GitHub has been doing some funny things with comments when diffs change.

File: scripts/pin_the_slow_way.py (hunk 9 above)
Comment (jakirkham) [inline scripts/pin_the_slow_way.py]: See comment above.

Comment (183amir) [inline scripts/pin_the_slow_way.py]: I see based on this: https://www.openssl.org/docs/faq.html#MISC8 you want to pin something like 'openssl >=1.0.2[any letter],<1.1' but how do you do that? This: 'openssl >=1.0.2*,<1.1' is not good.

Comment (183amir) [inline scripts/pin_the_slow_way.py]: Also based on this: https://abi-laboratory.pro/tracker/timeline/openssl/ what's written in the FAQ is not really true.

**Comment by pelson** [comment]
I think this has legs, and I'm supportive of it in general. Let's document the idea once and for-all in an [enhancement proposal](https://github.com/conda-forge/conda-forge-enhancement-proposals) with enough general detail that another packaging community could benefit from the knowledge it imparts, and use this PR as a reference implementation.

**Comment by 183amir** [comment]
@gqmelo or anybody else, wants to take care of the enhancement proposal. I really bad at writing these kind of things.

**Comment by gqmelo** [comment]
Sorry I forgot about this. Is the enhancement proposal still the way to go? I could try to write something.

But anyway I think this subject should be given some priority. Looking at the current ```pin_the_slow_way.py``` on master I've seen some new pins that were added since this PR and they got me worried. For example:

```python
'gmp': 'gmp >=5.0.1,<7', # 6.1.0
'mpfr': 'mpfr 3.*', # 3.1.5
'zlib': 'zlib 1.2.*', # 1.2.8
```

Using those pins on both ```build``` and ```run``` requirements allow conda to mix incompatible versions. 

Taking ```gmp``` as example, versions from  ```5.0.1``` up to ```6.1.2``` [are indeed backward compatible](https://abi-laboratory.pro/tracker/timeline/gmp/) but that does not mean that ```gmp >=5.0.1,<7``` should be used because this allows a package to be built with gmp ```6.1.0``` and run with ```5.0.1```. This is totally not gonna work as ```6.1.0``` has 34 symbols more than ```5.0.1```!

This shows some misunderstanding about API/ABI backward compatibility. When a specific version is said to be backward compatible it means that this version will work for a binary compiled on a previous version. It does not mean that all backward compatible versions could replace each other on any situation. The current scheme may be simpler, but IMHO is more error prone. At least the changes proposed on this PR will make more explicit that ```build``` and ```run``` requirements are different things.

**Comment by 183amir** [comment]
@gqmelo I think the best thing to do here is to go ahead with the proposal without explicitly pointing this PR as the implementation. Conda build 3 has some interesting pinning features. For implementation, I think it is best to wait for that and see what conda build 3 can offer us.

**Comment by jakirkham** [comment]
Would suggest it be discussed in PR ( https://github.com/conda-forge/staged-recipes/pull/2267 ). This is suppose to refactor out the pinnings for general use and be `conda-build` 3 compatible.

**Comment by 183amir** [comment]
I don't think with conda-build-3 this implementation is necessary anymore.

**Pull request closed without merging**